src/Sulu/Component/HttpCache/EventSubscriber/InvalidationSubscriber.php

```diff
@@ -57,25 +93,125 @@ public function __construct(
+    /**
+     * Invalidates all urls in the locale of the document of an already published document when it
+     * gets republished (eg on content change).
+     * This method is executed before the actual publishing of the document to avoid purging new urls.
+     *
+     * @param PublishEvent $event
+     */
+    public function invalidateDocumentBeforePublishing(PublishEvent $event)
+    {
+        $document = $event->getDocument();
+
+        if (!$document instanceof ResourceSegmentBehavior || !$document->getPublished()) {
+            // no need to invalidate the path to the document if it is not published currently
+            return;
+        }
+
+        // invalidate urls of the locale of the document
+        $locale = $this->documentInspector->getLocale($document);
+        foreach ($this->getLocaleUrls($document, $locale) as $url) {
+            $this->pathHandler->invalidatePath($url);
+        }
+    }
+
+    /**
+     * Invalidates all urls in the locale of the document of when a document gets unpublished.
+     * This method is executed before the actual unpublishing of the document because the document must still
+     * be published to gather the urls of the document.
+     *
+     * @param UnpublishEvent $event
+     */
+    public function invalidateDocumentBeforeUnpublishing(UnpublishEvent $event)
+    {
+        $document = $event->getDocument();
+
+        if (!$document instanceof ResourceSegmentBehavior || !$document->getPublished()) {
+            // no need to invalidate the path to the document if it is not published currently
+            return;
+        }
+
+        // invalidate urls of the locale of the document
+        $locale = $this->documentInspector->getLocale($document);
+        foreach ($this->getLocaleUrls($document, $locale) as $url) {
+            $this->pathHandler->invalidatePath($url);
+        }
+    }
+
+    /**
+     * Invalidates all urls in all locales of the document when a document gets removed.
+     * This method is executed before the actual removing of the document because the document must still
+     * exist to gather the urls of the document.
+     *
+     * @param RemoveEvent $event
+     */
+    public function invalidateDocumentBeforeRemoving(RemoveEvent $event)
     {
         $document = $event->getDocument();
 
-        if (!$document instanceof StructureBehavior) {
+        if (!$document instanceof ResourceSegmentBehavior) {
             return;
         }
 
-        $structureBridge = $this->structureManager->wrapStructure(
-            $this->documentInspector->getMetadata($document)->getAlias(),
-            $this->documentInspector->getStructureMetadata($document)
-        );
+        // invalidate urls of all locales of the document
+        $urls = [];
+        foreach ($this->documentInspector->getPublishedLocales($document) as $locale) {
+            $urls = array_merge($urls, $this->getLocaleUrls($document, $locale));
```

Comment:
Looks like the previous comment on this part got outdated... I think I would put another `foreach` in here or pass the stuff to another function. `array_merge` is not really necessary here, and it is also a quite unperformant function.